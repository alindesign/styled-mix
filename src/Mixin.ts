import { Properties, } from 'csstype';
import { css, FlattenInterpolation, ThemeProps } from 'styled-components';
import { castArray } from './utils/castArray';
import { nonNullableArray } from './utils/nonNullableArray';
import { dashCase } from './utils/dashCase';

export type MixinLookupType = string | string[];
export type MixinPropertyType = keyof Properties;

export type MixinPropsType = { [key: string]: any };

export type MixinLookupValue = null | { key: string; value: any; };

export type MixinOptions = {
    props?: MixinPropsType;
    property?: keyof Properties | string;
    lookup?: MixinLookupType;
    defaultValue?: any;
}

export class Mixin {
    private props: MixinPropsType = {};
    protected property?: keyof Properties | string;
    protected lookup?: MixinLookupType;
    protected defaultValue: any = null

    public constructor({
        props,
        property,
        lookup,
        defaultValue
    }: MixinOptions = {}) {
        this.props = props ?? {};
        this.property = property ?? this.property;
        this.lookup = lookup ?? this.lookup;
        this.defaultValue = defaultValue ?? this.defaultValue;

        this.mixin = this.mixin.bind(this);
    }


    public lookupValue(): MixinLookupValue {
        const lookup = nonNullableArray<string>(castArray(this.lookup).map((prop) => {
            if (prop) {
                return prop.replace('@', this.property);
            }

            return prop;
        }));

        if (lookup.length === 0) {
            return null;
        }

        const result = lookup.find((lookupProp: string) => (!!this.props[lookupProp] || this.props[lookupProp] === 0));

        if (!result) {
            return null;
        }

        return {
            key: result,
            value: this.props[result]
        }
    }

    public build(props?: MixinPropsType): FlattenInterpolation<ThemeProps<any>> {
        this.extendProps(props);

        return css`${dashCase(this.property)}: ${this.mixin};`;
    }

    private mixin(cssProps: any) {

        this.extendProps(cssProps);

        let result = this.lookupValue();
        if (!result) {
            result = {
                key: '$',
                value: this.defaultValue
            };
        }

        return this.transform(result.value);
    }

    protected transform(value: any): any {
        if (typeof value === 'function') {
            return value({ ...this });
        }

        return `${value}`.trim();
    }

    public extendProps(props?: MixinPropsType) {
        this.props = { ...(this.props ?? {}), ...(props ?? {}) };
        return this;
    }

    toString(props?: MixinPropsType) {
        return this.build(props)
    }
}
