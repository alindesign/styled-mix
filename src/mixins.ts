import { Mixin } from "./Mixin";
import { css } from "styled-components";

type MixinClass = new (...args: any) => Mixin;

export const mixins = (...mixins: (MixinClass | Mixin)[]) => (
    css`${(props) => mixins.map((mixinDeclaration: MixinClass | Mixin) => {
        const mixin = mixinDeclaration instanceof Mixin ? mixinDeclaration : new mixinDeclaration({ props });

        return mixin.build(props);
    })}`
);
