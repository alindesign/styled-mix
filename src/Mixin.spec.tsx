import styled from 'styled-components'
import renderer from 'react-test-renderer';
import 'jest-styled-components'

import { Mixin } from "./Mixin";
import React from 'react';
import { mixins } from './mixins';

describe('styled-mixins', () => {
    describe('Mixin', () => {
        it('should correctly lookup for the properties list', () => {
            const opts = {
                property: 'color',
                lookup: ['a', 'b', 'c', 'd', 'e']
            };

            const mix = new Mixin(opts);
            mix.extendProps({ c: 3 });

            expect(mix.lookupValue()).not.toBeNull();
            expect(mix.lookupValue()).toMatchObject({ key: 'c', value: 3 });

            const mix2 = new Mixin(opts);
            mix2.extendProps({ a: undefined, b: null, c: '', d: 0, e: 1 });

            expect(mix2.lookupValue()).not.toBeNull();
            expect(mix2.lookupValue()).toMatchObject({ key: 'd', value: 0 });

            const mix3 = new Mixin(opts);

            expect(mix3.lookupValue()).toBeNull();
        });

        it('should correctly render a simple color mixin', () => {
            const colorMixin = new Mixin({ property: 'color', lookup: ['c', 'color'], defaultValue: '#222' });
            const Button = styled.button<any>`${ mixins(colorMixin) }`;

            const tree1 = renderer.create(<Button />).toJSON()
            expect(tree1).toHaveStyleRule('color', '#222');

            const tree2 = renderer.create(<Button color={'#fff'} />).toJSON()
            expect(tree2).toHaveStyleRule('color', '#fff');

            const tree3 = renderer.create(<Button c={'#333'} />).toJSON()
            expect(tree3).toHaveStyleRule('color', '#333');
        });


        it('should correctly render advanced mixin', () => {
            class MyMixin extends Mixin {
                protected property = 'color';
                protected lookup = ['c', 'color'];
            }

            class MarginMixin extends Mixin {
                protected property = 'margin';
                protected lookup = ['m', 'margin'];

                transform(value: any) {
                    if (typeof value === "number") {
                        return `${value}px`;
                    }

                    return value;
                }
            }

            const Button = styled.button<any>`${mixins(MyMixin, MarginMixin)}`;

            const tree1 = renderer.create(<Button />).toJSON()
            expect(tree1).toMatchSnapshot();

            expect(tree1).not.toHaveStyleRule('color');
            expect(tree1).not.toHaveStyleRule('margin');

            const tree2 = renderer.create(<Button color={'#fff'} />).toJSON()
            expect(tree2).toHaveStyleRule('color', '#fff');

            const tree3 = renderer.create(<Button c={'#333'} />).toJSON()
            expect(tree3).toHaveStyleRule('color', '#333');

            const tree4 = renderer.create(<Button margin={8} />).toJSON()
            expect(tree4).toHaveStyleRule('margin', '8px');

            const tree5 = renderer.create(<Button margin={'8px'} />).toJSON()
            expect(tree5).toHaveStyleRule('margin', '8px');

            const tree6 = renderer.create(<Button m={8} />).toJSON()
            expect(tree6).toHaveStyleRule('margin', '8px');

            const tree7 = renderer.create(<Button m={'8px'} />).toJSON()
            expect(tree7).toHaveStyleRule('margin', '8px');
        });
    });
});
