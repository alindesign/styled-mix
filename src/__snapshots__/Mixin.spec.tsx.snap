// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled-mixins Mixin should correctly render advanced mixin 1`] = `
.c0 {
  color: null;
}

<button
  className="c0"
/>
`;
